### PR TITLE
Proposal: Attributes as properties

### PR DESCRIPTION
**Pull request by subzey**
Hey! Me again :smile: 

Not that it's a ready-to-merge MR, rather a suggestion.

What if instead of syncing manually internal properties and attribute values we could just use the attributes as a state. It has several benefits:

- No need for handling attr change and property change separately
- Correct `removeAttribute()` handling out of the box
- Better separation of the basic abstract classes
- 0.02 Kb less in `size-limit`

----

The idea is using `attributeChangedCallback` for handling all the changes. _In fact, `.attributeChangedCallback()` and `[update]()` now can be merged into one function, but would clutter the diff._

- The attributes are *public*, *unsanitized* values. Meaning, the user can set any value they want. That's really how HTML usually works.
- The properties setters are *public* and *unsanitized*. Again, it mimics the DOM behavior: you can set `el.zIndex = 'totally garbage'` and it would just handle it as `0`
- The property getters are *public* and **sanitized**. They are also used for the internal machinery (this part is unchanged).

In short: 

```ts
burger.size = 'some invalid value';
burger.size === 'md'; // sanitized: the default value;
burger.getAttribute('size'); // 'some invalid value' (unsanitized)
```

----

I've run the tests and size-limit so this time hopefully I won't drop a broken code in my PR

### CODE REVIEW

**Comment by web-padawan** [comment]
Hi, thanks for putting this together!

I was thinking about keeping attributes and properties in sync (either by [reflecting](https://github.com/Polymer/lit-element/blob/2de92b532b8adad540b91f208063b2e15777029f/src/lib/updating-element.ts#L858-L866) or as you suggested).
For this component it might make sense because all the properties are primitive (no objects / arrays).

Looking at your code I see the following shortcomings:

## 1. Property value change workflow is counter-intuitive. 

Consider the following case, which would be in effect if this change is accepted:

```js
burger.distance = null;
console.log(burger.distance); // 'md'
```

Currently `null` is accepted and returned (although in this case value corresponding to `md` is still used internally).
Another example where `null` actually makes sense is `easing` property (where it disables animation).

## 2. Reflecting to attribute is not opt-in

Currently it is possible to extend `Burger` and customize it so that it would reflect property values to attributes.
With the new implementation it will be always enabled and will not be possible to opt out.

While in this simple case `getAttribute` and `setAttribute` should be cheap, it might be a concern for someone.
There are reasons why libraries like LitElement make reflecting to attribute opt-in.

Looking at the [Custom Elements Best Practices](https://developers.google.com/web/fundamentals/web-components/best-practices#aim-to-keep-primitive-data-attributes-and-properties-in-sync,-reflecting-from-property-to-attribute,-and-vice-versa.), I see the reason behind reflecting but IMO it's a bit opinionated.

## 3. Consistency

Personally I'm mostly concerned about consistency. As we know, native `input` does not sync neither `value` nor `checked` properties to attributes. On the other hand, `<details>` element **does** reflect `open` to attribute.

My other project [vanilla-colorful](https://github.com/web-padawan/vanilla-colorful) does not reflect to attributes because it supports different elements with `color` property as string or object, and reflecting in some cases but not in the others is not something I want to support.

## Summary

Personally I don't see a big problem with the current implementation and hopefully I explained the current approach.

Could you please elaborate on the problem with `removeAttribute`, do you think it makes using the component confusing?

**Comment by subzey** [comment]
First of all, thank you for your review!

A user may want to read the *effective* size of the burger and they would check it like this:

```
if (burger.size === 32 || burger.size === null)
```

That looks like an abstraction leak. But really... I suppose, in the most cases the library user won't ever read or write anything but the `.pressed` in the run-time, so there's no big difference for them.

----

Currently, the resetting of the attribute would just set null and overwrite the last set property value.

```
$0.size; // 32
$0.size = 22; $0.size; // 22
$0.setAttribute('size', 42); $0.size; // 42
$0.removeAttribute('size'); $0.size; // null
```

Getting `null` feels weird, but on the other hand... Isn't `null` the desired value after this operation? Or should it be the last set `22`? Or the default `32`? 

But that's nothing compared to the builtin `input.checked` value handling you mentioned! The `setAttribute('checked') / removeAttribute('checked')` works fine but only until you set the `.checked` property. After that the attribute modification is ignored. :exploding_head: 

If the attributes are reflected, the different states may be styled differently. AFAIK, the new one `<details>` works just like that. That may be the main reason `input.value` (as well as `element.nonce`) is not reflected.

----

Really, I have no strong preference or arguments of how is is it right or wrong. It's just reflecting the attributes and returning the "actual" values from getters subjectively feels more straightforward. But again, it's up to you to decide!

**Comment by web-padawan** [comment]
After some more thinking, I do see a benefit in reflecting `pressed` property for styling. Submitted #6 to handle that.

**Comment by subzey** [comment]
#6 Looks great!

Reflecting other props as attrs most probably won't change anyone's life to the better, so I'm closing this MR

**Comment by web-padawan** [comment]
Once again, thank you for raising this question!
